crates/swc_ecma_transforms_base/src/ident_scope.rs

```diff
@@ -0,0 +1,110 @@
+use std::{cell::RefCell, rc::Rc};
+
+use is_macro::Is;
+use swc_atoms::JsWord;
+use swc_common::{collections::AHashMap, SyntaxContext};
+use swc_ecma_ast::{AssignPatProp, Expr, Ident, KeyValuePatProp, PropName, VarDecl, VarDeclKind};
+use swc_ecma_visit::{
+    as_folder, noop_visit_mut_type, noop_visit_type, Fold, Visit, VisitMut, VisitMutWith, VisitWith,
+};
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+pub struct IdentScope {
+    sym: JsWord,
+    ctxt: SyntaxContext,
+}
+
+impl From<(JsWord, SyntaxContext)> for IdentScope {
+    fn from((sym, ctxt): (JsWord, SyntaxContext)) -> Self {
+        IdentScope { sym, ctxt }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Is, Hash)]
+pub enum IdentScopeKind {
+    /// `var`
+    Var,
+    /// `let` or `const`
+    Block,
+    /// `var` but should be block scoped
+    BlockVar,
+    Unresolved,
+}
+
+impl Default for IdentScopeKind {
+    fn default() -> Self {
+        IdentScopeKind::Unresolved
+    }
+}
+
+impl From<VarDeclKind> for IdentScopeKind {
+    fn from(value: VarDeclKind) -> Self {
+        if value == VarDeclKind::Var {
+            Self::Var
+        } else {
+            Self::Block
+        }
+    }
+}
+
+pub type IdentScopeRecord = Rc<RefCell<AHashMap<IdentScope, IdentScopeKind>>>;
+
+pub fn ident_scope_collector(ident_scope_record: IdentScopeRecord) -> impl Fold {
+    as_folder(IdentScopeCollector { ident_scope_record })
+}
+
+struct IdentScopeCollector {
+    ident_scope_record: IdentScopeRecord,
+}
+
+impl VisitMut for IdentScopeCollector {
+    noop_visit_mut_type!();
+
+    fn visit_mut_var_decl(&mut self, var: &mut VarDecl) {
+        var.visit_mut_children_with(self);
+
+        let mut pat_visitor = PatVisitor {
+            ident_scope_record: self.ident_scope_record.clone(),
+            ident_scope_kind: var.kind.into(),
+        };
+
+        for decl in var.decls.iter() {
+            decl.name.visit_with(&mut pat_visitor);
+        }
+    }
+}
+
+struct PatVisitor {
+    ident_scope_record: IdentScopeRecord,
+    ident_scope_kind: IdentScopeKind,
+}
+
+impl Visit for PatVisitor {
+    noop_visit_type!();
+
+    /// No-op (we don't care about expressions)
+    fn visit_expr(&mut self, _: &Expr) {}
+
+    fn visit_ident(&mut self, ident: &Ident) {
+        self.ident_scope_record
+            .borrow_mut()
+            .entry(IdentScope {
+                sym: ident.sym.clone(),
+                ctxt: ident.span.ctxt(),
+            })
+            .or_insert_with(|| self.ident_scope_kind);
+    }
+
```

Comment:
We need `visit_object_pat_prop` because it creates a binding.

Reply:
https://github.com/swc-project/swc/blob/2ea81db19fb5a36ecd949c06fa242525d62c8997/crates/swc_ecma_ast/src/pat.rs#L116-L125

According to the definition of `ObjectPatProp`, I think it is only necessary to handle with `KeyValuePatProp` and `AssignPatProp`.

See:

https://github.com/swc-project/swc/blob/3afcaba887539e2cce40f5cdd4fbce0511c77ff0/crates/swc_ecma_transforms_base/src/ident_scope.rs#L101-L109

Reply:
Oh, I missed it. My bad, sorry.